lib/std/debug.zig

```diff
@@ -27,185 +42,475 @@ pub const runtime_safety = switch (builtin.mode) {
+    config.captureStackTraceFrom
+else
+    default_config.captureStackTraceFrom;
+
+/// Slightly different name than in config to avoid redefinition in default.
+const getPanicStageTL = if (@hasDecl(config, "getPanicStage"))
+    config.getPanicStage
+else
+    default_config.getPanicStage;
+
+/// Slightly different name than in config to avoid redefinition in default.
+const sleepForev = if (@hasDecl(config, "sleepForever"))
+    config.sleepForever
+else
+    default_config.sleepForever;
+
+/// Slightly different name than in config to avoid redefinition in default.
+const panicTerm = if (@hasDecl(config, "panicTerminate"))
+    config.panicTerminate
+else
+    default_config.panicTerminate;
+
+var print_mutex = std.Thread.Mutex{};
+
+pub fn getPrintMutex() *std.Thread.Mutex {
+    return &print_mutex;
+}
+
+pub const getStderrMutex = @compileError("This was renamed to getPrintMutex because " ++
+    "the writer used by `debug` can be overriden.");
 
 /// Deprecated. Use `std.log` functions for logging or `std.debug.print` for
 /// "printf debugging".
 pub const warn = print;
 
-/// Print to stderr, unbuffered, and silently returning on failure. Intended
+/// Print to log writer, unbuffered, and silently returning on failure. Intended
 /// for use in "printf debugging." Use `std.log` functions for proper logging.
 pub fn print(comptime fmt: []const u8, args: anytype) void {
-    const held = stderr_mutex.acquire();
+    const held = print_mutex.acquire();
     defer held.release();
-    const stderr = io.getStdErr().writer();
-    nosuspend stderr.print(fmt, args) catch return;
-}
-
-pub fn getStderrMutex() *std.Thread.Mutex {
-    return &stderr_mutex;
+    const writer = getWrit();
+    nosuspend writer.print(fmt, args) catch {};
 }
 
 /// TODO multithreaded awareness
-var self_debug_info: ?DebugInfo = null;
+var sym_map: ?SymMap = null;
 
-pub fn getSelfDebugInfo() !*DebugInfo {
-    if (self_debug_info) |*info| {
+pub fn getSymbolMap() !*SymMap {
+    if (sym_map) |*info| {
         return info;
     } else {
-        self_debug_info = try openSelfDebugInfo(getDebugInfoAllocator());
-        return &self_debug_info.?;
+        sym_map = try SymMap.init(getDebugInfoAllocator());
+        return &sym_map.?;
     }
 }
 
-pub fn detectTTYConfig() TTY.Config {
-    var bytes: [128]u8 = undefined;
-    const allocator = &std.heap.FixedBufferAllocator.init(bytes[0..]).allocator;
-    if (process.getEnvVarOwned(allocator, "ZIG_DEBUG_COLOR")) |_| {
-        return .escape_codes;
-    } else |_| {
-        const stderr_file = io.getStdErr();
-        if (stderr_file.supportsAnsiEscapeCodes()) {
-            return .escape_codes;
-        } else if (builtin.os.tag == .windows and stderr_file.isTty()) {
-            return .windows_api;
-        } else {
-            return .no_color;
-        }
+const is_stripped = using_default_symbol_map and builtin.strip_debug_info;
+
+fn getMappingForDump(writer: anytype) ?*SymMap {
+    if (is_stripped) {
```

Comment:
This is actually wrong as we may not have debug symbols for the Zig binary but have them for other binaries/libs we're called from.

(That's why the symbol mapper is iterating over all the loaded objects, that and being able to handle position-independent code)

Reply:
Unfortunately, removing this check currently breaks the wasm build (maybe just test?). But yeah, this is certainly wrong and I will think about how this should be solved.

Reply:
Ugh, looks like my investigation into how `wasm` needs to work has just revealed more bugs in the implementation.

At the moment, the type of ModuleDebugInfo is DwarfInfo in the other case (not a supported os). This makes no sense because ModuleDebugInfo must have a `addressToSymbol` method. Also, `getModuleForAddress` makes no sense in the other case as well. I think `getModuleForAddress` should be a function on the ModuleDebugInfo itself, at the moment, it is a totally separate function which just assumes the type of the ModuleDebugInfo (which is very strange).

Reply:
Also, `lookupModuleHaiku` just panics with a TODO...

Reply:
To improve my sanity I going to fix all of this...